File: 1298-reverse-substrings-between-each-pair-of-parentheses/1298-reverse-substrings-between-each-pair-of-parentheses.java
class Solution {
    public String reverseParentheses(String s) {
        Stack<Integer> st = new Stack<>();
        StringBuilder ans = new StringBuilder();
        for(char c : s.toCharArray()){
            if(c == '(') st.push(ans.length());
            else if(c == ')'){
                int start = st.pop();
                reverse(ans, start, ans.length()-1);
            }
            else ans.append(c);
        }
        return ans.toString();
    }
    private void reverse(StringBuilder sb, int l, int r){
        while(l < r){
            char temp = sb.charAt(l);
            sb.setCharAt(l++, sb.charAt(r));
            sb.setCharAt(r--, temp);
        }
    }
}

File: 1298-reverse-substrings-between-each-pair-of-parentheses/1298-reverse-substrings-between-each-pair-of-parentheses.cpp
class Solution {
public:
    string reverseParentheses(string s) {
        stack<int> openParenthesesIndices;
        string ans;
        for(char c : s){
            if(c == '('){
                openParenthesesIndices.push(ans.length());
            }
            else if(c == ')'){
                int start = openParenthesesIndices.top();
                openParenthesesIndices.pop();
                reverse(ans.begin()+start, ans.end());
            }
            else ans += c;
        }
        return ans;
    }
};